sklearn/decomposition/_nmf.py

```diff
@@ -1188,6 +1213,14 @@ def _check_params(self, X):
     def _check_w_h(self, X, W, H, update_H):
         """Check W and H, or initialize them."""
         n_samples, n_features = X.shape
+        if self.n_components == "auto":
+            if H is not None:
+                self._n_components = _num_samples(H)
+            elif W is not None:
+                self._n_components = _num_features(W)
```

Comment:
Small note about this branch:

if `H is None and W is not None` we now set `self._n_components = _num_features(W)` instead of `self._n_components = X.shape[1]`
#### Situation before to the PR
If you have `init = "custom"` ==> Error because `W` is not set
If you have `init != "custom"` ==> A new `W` is created with the shape `n_samples, X.shape[1]`. This is misleading because `X.shape[1]` can be different from `_num_features(W)`
see https://github.com/scikit-learn/scikit-learn/blob/9cbcc1f205e8be4dad1f383239e98381abb28bd0/sklearn/decomposition/_nmf.py#L1208-L1216

#### Situation after the PR
If you have `init = "custom"` [Same behaviour] ==> Error because `W` is not set
If you have `init != "custom"` [Changed behaviour]==> a new `W` is created with but now the shape of `W` will be the same as the shape of the `W` passed as argument

Reply:
the test was added in d8b6bf84ac93ef0b9fbd84cef45a0d12c0819705

Reply:
> If you have init != "custom" [Changed behaviour]==> a new W is created with but now the shape of W will be the same as the shape of the W passed as argument

I don't think this is a desirable behavior. It's also a breaking change. In that case we should completely ignore W and fallback to X.shape[1].

It requires a little bit of refactoring because it means that we can't compute n_components before checking in which situation we are. I directly pushed some changes to fix the behavior. The benefit is that we no longer have to rely on ``_num_samples`` and ``_num_features`` :)